crates/ruma-client-api/src/sync/sync_events.rs

```diff
@@ -62,3 +73,174 @@ impl DeviceLists {
+pub enum StrippedState {
+    /// A stripped state event.
+    Stripped(AnyStrippedStateEvent),
+
+    /// A timeline state event.
+    #[cfg(feature = "unstable-msc4311")]
+    Timeline(AnyStateEvent),
+}
+
+impl StrippedState {
+    /// Returns the `type` of this event.
+    pub fn event_type(&self) -> StateEventType {
+        match self {
+            Self::Stripped(event) => event.event_type(),
+            #[cfg(feature = "unstable-msc4311")]
+            Self::Timeline(event) => event.event_type(),
+        }
+    }
+
+    /// Returns this event's `sender` field.
+    pub fn sender(&self) -> &UserId {
+        match self {
+            Self::Stripped(event) => event.sender(),
+            #[cfg(feature = "unstable-msc4311")]
+            Self::Timeline(event) => event.sender(),
+        }
+    }
+
+    /// Returns this event's `state_key` field.
+    pub fn state_key(&self) -> &str {
+        match self {
+            Self::Stripped(event) => event.state_key(),
+            #[cfg(feature = "unstable-msc4311")]
+            Self::Timeline(event) => event.state_key(),
+        }
+    }
+}
+
+impl<'de> Deserialize<'de> for StrippedState {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: ::serde::Deserializer<'de>,
+    {
+        let json = Box::<RawJsonValue>::deserialize(deserializer)?;
+
+        #[cfg(feature = "unstable-msc4311")]
+        {
+            use serde::de;
+
+            #[derive(Deserialize)]
+            struct PotentialTimelineEventDeHelper {
+                event_id: Option<de::IgnoredAny>,
+                origin_server_ts: Option<de::IgnoredAny>,
+                room_id: Option<de::IgnoredAny>,
+            }
+
+            let PotentialTimelineEventDeHelper { event_id, origin_server_ts, room_id } =
+                from_raw_json_value(&json)?;
+
+            if event_id.is_some() && origin_server_ts.is_some() && room_id.is_some() {
+                return from_raw_json_value(&json).map(Self::Timeline);
+            }
+        }
+
+        from_raw_json_value(&json).map(Self::Stripped)
+    }
+}
+
+impl From<AnyStrippedStateEvent> for StrippedState {
+    fn from(value: AnyStrippedStateEvent) -> Self {
+        Self::Stripped(value)
+    }
+}
+
+#[cfg(feature = "unstable-msc4311")]
+impl From<AnyStateEvent> for StrippedState {
+    fn from(value: AnyStateEvent) -> Self {
+        Self::Timeline(value)
+    }
+}
+
+impl JsonCastable<StrippedState> for AnyStrippedStateEvent {}
+
+#[cfg(feature = "unstable-msc4311")]
+impl JsonCastable<StrippedState> for AnyStateEvent {}
+
+impl<C> JsonCastable<StrippedState> for StrippedStateEvent<C> where
+    C: PossiblyRedactedStateEventContent
+{
+}
+
+#[cfg(feature = "unstable-msc4311")]
+impl<C> JsonCastable<StrippedState> for StateEvent<C>
+where
+    C: StaticStateEventContent + RedactContent,
+    <C as RedactContent>::Redacted: RedactedStateEventContent,
+{
+}
+
+impl JsonCastable<JsonObject> for StrippedState {}
```

Comment:
We should have all the same `JsonCastable` implementations as in #2170 and no need to put them behind cargo features, because in the worst case they will all successfully deserialize to the `Stripped` variant.